identity-iota/src/tangle/client.rs

```diff
@@ -97,6 +105,27 @@ impl Client {
       .map(|message| Receipt::new(self.network.clone(), message))
   }
 
+  /// Publishes arbitrary JSON data to the specified index on the Tangle.
+  /// Retries (promotes or reattaches) the message until it’s included (referenced by a milestone).
+  /// Default interval is 5 seconds and max attempts is 40.
+  pub async fn publish_json_with_retry<T: ToJson>(
+    &self,
+    index: &str,
+    data: &T,
+    interval: Option<u64>,
+    max_attempts: Option<u64>,
+  ) -> Result<Receipt> {
+    let receipt = self.publish_json(index, data).await?;
+    let reattached_messages = self
+      .client
+      .retry_until_included(receipt.message_id(), interval, max_attempts)
+      .await?;
+    match reattached_messages.first() {
+      Some((_, message)) => Ok(Receipt::new(self.network.clone(), message.clone())),
+      None => Ok(receipt),
+    }
```

Comment:
Missed this initially: do we really need to clone the entire message which includes the payload, can we extract it from the list instead?

Reply:
E.g.

```suggestion
    match reattached_messages.into_iter().next() {
      Some((_, message)) => Ok(Receipt::new(self.network.clone(), message)),
      None => Ok(receipt),
    }
```